src/hivpy/hiv_testing.py

```diff
@@ -145,67 +194,77 @@ def update_anc_hiv_testing(self, pop, time_step):
-        """
-        if pop.circumcision.circ_after_test:
-            # select uncircumcised men tested last timestep
-            tested_uncirc_male_pop = pop.get_sub_pop([(col.SEX, op.eq, SexType.Male),
-                                                      (col.CIRCUMCISED, op.eq, False),
-                                                      (col.HIV_DIAGNOSED, op.eq, False),
-                                                      (col.LAST_TEST_DATE, op.eq, pop.date - time_step),
-                                                      (col.HARD_REACH, op.eq, False),
-                                                      (col.AGE, op.le, pop.circumcision.max_vmmc_age)])
-            # continue if eligible men are present this timestep
-            if len(tested_uncirc_male_pop) > 0:
-                # calculate post-test vmmc outcomes
-                r = rng.uniform(size=len(tested_uncirc_male_pop))
-                circumcision = r < pop.circumcision.prob_circ_after_test
-                # assign outcomes
-                pop.set_present_variable(col.CIRCUMCISED, circumcision, tested_uncirc_male_pop)
-                pop.set_present_variable(col.VMMC, circumcision, tested_uncirc_male_pop)
-
-    def update_post_vmmc_testing(self, pop):
-        """
-        Update HIV testing status after VMMC.
+            self.update_sub_pop_test_mark(pop, post_delivery_pop, self.prob_test_postdel)
+
+    def update_sub_pop_test_mark(self, pop, sub_pop, prob_test):
         """
-        # those that just got circumcised and weren't tested last time step get tested now
-        just_tested = pop.get_sub_pop(AND(COND(col.CIRCUMCISION_DATE, op.eq, pop.date),
-                                          OR(COND(col.LAST_TEST_DATE, op.lt, pop.date - timedelta(days=90)),
-                                             COND(col.LAST_TEST_DATE, op.eq, None))))
-        # correctly set up related columns
-        if len(just_tested) > 0:
-            pop.set_present_variable(col.EVER_TESTED, True, just_tested)
-            pop.set_present_variable(col.LAST_TEST_DATE, pop.date, just_tested)
-            # "reset" dummy partner columns
-            pop.set_present_variable(col.NSTP_LAST_TEST, 0, just_tested)
-            pop.set_present_variable(col.NP_LAST_TEST, 0, just_tested)
+        Update the test mark of a sub-population based on a given probability.
+        """
+        if len(sub_pop) > 0:
+            # mark people for testing
+            r = rng.uniform(size=len(sub_pop))
+            marked = r < prob_test
+            # set outcomes
+            pop.set_present_variable(col.TEST_MARK, True, sub_pop=pop.apply_bool_mask(marked, sub_pop))
+
+    def test_mark_general_pop(self, pop):
+        """
+        Mark general population to undergo testing this time step.
+        """
+        # testing occurs after a certain year if there is no covid disruption
+        if ((pop.date.year >= self.date_start_testing)
```

Comment:
SAS code features `caldate{t} ge (date_start_testing+5.5)` for this section, should this date be pushed later?

Reply:
@andrew-phillips-1 What is the difference between the use of `date_start_testing` and `date_start_testing+5.5` in the SAS code? Crucially, I don't think we can include the +5.5 in `date_start_testing` as noted above because it would change the pattern of increasing testing which needs to start at the earlier date in the SAS code:
```
if date_start_testing lt caldate{t} le 2015  then do; 
	test_rate_who4 = min(0.9,test_rate_who4*incr_test_rate_sympt);  
	test_rate_tb  = min(0.8,test_rate_tb*incr_test_rate_sympt);  
	test_rate_non_tb_who3 = min(0.7,test_rate_non_tb_who3*incr_test_rate_sympt); * 0.7 mar19;
* testing for hiv for a person with non_hiv_tb (i.e. who was hiv negative in last period) ;  * update_24_4_21;
	*This is intended for people who did not have HIV 3 months ago;																																																
	if caldate{t} - date_last_non_hiv_tb = 0.25 and tested ne 1 then do;   * ts1m - dependent on time step ;
		e=rand('uniform'); 
		if e < test_rate_tb then do;
			tested=1; 
			if ever_tested ne 1 then date1test=caldate{t}; ever_tested=1; dt_last_test=caldate{t}; 
			np_lasttest=0; newp_lasttest_tested_this_per=newp_lasttest; newp_lasttest=0;
		end;
	end;
end;
```

Reply:
Thanks for pointing this out Michael @mmcleod89 .  The point about the 5.5 is that we wanted the rate of testing to start increasing from 5.5 years after testing introduction in 2003.5.  So having been constant until 2009 the rate starts to increase until date_test_rate_plateau.

But I can see that in this line:
rate_1sttest = initial_rate_1sttest + (min(caldate{t},date_test_rate_plateau)-(date_start_testing+5.5))*an_lin_incr_test;

we need to ensure that the second term is not negative, which I think it currently can be.

We don't need the loop starting: if date_start_testing lt caldate{t} le 2015  then do; 

because we don't change the date of start of testing and it is well before 2015.

But I'm not sure if this is answering your question ?

Reply:
We need:

- date_testing_introduced
- date for general population testing growth 
- date for general population testing plateau (date_general_testing_plateau ?)
- date for specific testing (symptomatic, ANC etc.) testing plateau